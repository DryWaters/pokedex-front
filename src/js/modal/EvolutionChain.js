import React from 'react';
import PropTypes from 'prop-types';

import Evolution from './Evolution';

import '../../styles/modal/EvolutionChain.css';

const propTypes = {
  getPokemonDetails: PropTypes.func.isRequired,
  evolutions: PropTypes.shape({
    1: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        types: PropTypes.arrayOf(PropTypes.string).isRequired,
        image_path: PropTypes.string.isRequired,
      }).isRequired
    ),
    2: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        types: PropTypes.arrayOf(PropTypes.string).isRequired,
        image_path: PropTypes.string.isRequired,
      }).isRequired
    ),
    3: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        types: PropTypes.arrayOf(PropTypes.string).isRequired,
        image_path: PropTypes.string.isRequired,
      }).isRequired
    ),
  }).isRequired,
};

class EvolutionChain extends React.Component {
  mapArrayToEvolutions(arr) {
    return arr.map((evolution, index) => {
      return <Evolution key={index}
                        getPokemonDetails={this.props.getPokemonDetails}
                        {...arr[index]}
             />;
    });
  }

  renderEvolutions() {
    const firstGroup = this.props.evolutions[1] === null ? null :  this.mapArrayToEvolutions(this.props.evolutions[1]);
    const secondGroup = this.props.evolutions[2] === null ? null : this.mapArrayToEvolutions(this.props.evolutions[2]);
    const thirdGroup = this.props.evolutions[3] === null ? null : this.mapArrayToEvolutions(this.props.evolutions[3]);

    return (
      <div className="evolution-chain-container">
        <div className="evolution-chain-item first">{firstGroup}</div>
        <div className="evolution-chain-item second">{secondGroup}</div>
        <div className="evolution-chain-item third">{thirdGroup}</div>
      </div>
    );
  }

  render () {
    return (
      <div className="evolution-chain-background">
        <h3>{this.props.evolutions[1] ? 'Evolutions' : 'This pokemon doesn\'t evolve'}</h3>
        {this.props.evolutions[1] ? this.renderEvolutions() : null}
      </div>
    );
  }
}

EvolutionChain.propTypes = propTypes;

export default EvolutionChain;
