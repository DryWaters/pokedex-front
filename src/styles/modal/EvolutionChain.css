.evolution-chain-container {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 40px 0;
  background-color: #FFFFC6;
}

.evolution-chain-item {
  display: flex;
  flex-direction: column;
  margin: 0 40px;
  /*flex-wrap: wrap;
  max-height: 1000px;*/
}

.second .evolution-container::before, .third .evolution-container::before {
  content: "\203A";
  font-size: 80px;
  position: relative;
  top: 130px;
  left: -100px;
}

.second .evolution-container, .third .evolution-container {
  margin-top: -120px; /* Offset the `top` from ::before selectors; */
}
