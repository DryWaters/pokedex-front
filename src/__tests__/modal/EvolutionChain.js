import React from 'react';
import { shallow } from 'enzyme';

import EvolutionChain from '../../js/modal/EvolutionChain';

import ExampleJSON from '../../../local_data/example_detail.json';
import ExampleJSON2 from '../../../local_data/example_detail_2.json';
import ExampleJSON3 from '../../../local_data/example_detail_3.json';

describe('EvolutionChain Component', () => {
  describe('when there are evolutions', () => {
    let props;
    beforeEach(() => {
      props = {
        getPokemonDetails: (() => {}),
        evolutions: ExampleJSON.evolutions,
      };
    });

    it('should render EvolutionChain component without crashing', () => {
      shallow(<EvolutionChain {...props} />);
    });

    it('should match EvolutionChain component snapshot', () => {
      const wrapper = shallow(<EvolutionChain {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should have an evolution chain within it', () => {
      const wrapper = shallow(<EvolutionChain {...props} />);
      expect(wrapper.find('.evolution-chain-container')).toHaveLength(1);
    });

    describe('Evolution components in an evolution chain', () => {
      it('should have 3 Evolution components for the Butterfree chain', () => {
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('Evolution')).toHaveLength(3);
      });

      it('should have 5 Evolution components for the Cascoon chain', () => {
        props.evolutions = ExampleJSON2.evolutions;
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('Evolution')).toHaveLength(5);
      });

      it('should have 2 Evolution components for the Rattata chain', () => {
        props.evolutions = ExampleJSON3.evolutions;
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('Evolution')).toHaveLength(2);
      });
    });

    describe('evolution items in an evolution chain', () => {
      it('should have 3 evolution items for the Butterfree chain', () => {
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('.evolution-chain-item')).toHaveLength(3);
      });

      it('should have 3 evolution items for the Cascoon chain', () => {
        props.evolutions = ExampleJSON2.evolutions;
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('.evolution-chain-item')).toHaveLength(3);
      });

      it('should have 2 evolution items for the Rattata chain', () => {
        props.evolutions = ExampleJSON3.evolutions;
        const wrapper = shallow(<EvolutionChain {...props} />);
        expect(wrapper.find('.evolution-chain-item')).toHaveLength(2);
      });
    });

  });

  describe('when there are no evolutions', () => {
    let props;
    beforeEach(() => {
      props = {
        getPokemonDetails: (() => {}),
        evolutions: {
          1: null,
          2: null,
          3: null,
        },
      };
    });

    it('should render EvolutionChain component without crashing', () => {
      shallow(<EvolutionChain {...props} />);
    });

    it('should match EvolutionChain component snapshot', () => {
      const wrapper = shallow(<EvolutionChain {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should not have an evolution chain within it', () => {
      const wrapper = shallow(<EvolutionChain {...props} />);
      expect(wrapper.find('.evolution-chain-container')).toHaveLength(0);
    });
  });
});
