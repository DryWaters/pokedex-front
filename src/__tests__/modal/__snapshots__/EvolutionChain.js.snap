// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvolutionChain Component when there are evolutions should match EvolutionChain component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EvolutionChain
    evolutions={
        Object {
            "1": Array [
              Object {
                "id": 10,
                "image_path": "/sprites/pokemon/small/10.png",
                "name": "Caterpie",
                "types": Array [
                  "bug",
                ],
              },
            ],
            "2": Array [
              Object {
                "id": 11,
                "image_path": "/sprites/pokemon/small/11.png",
                "name": "Metapod",
                "types": Array [
                  "bug",
                ],
              },
            ],
            "3": Array [
              Object {
                "id": 12,
                "image_path": "/sprites/pokemon/small/12.png",
                "name": "Butterfree",
                "types": Array [
                  "bug",
                  "flying",
                ],
              },
            ],
          }
    }
    getPokemonDetails={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Evolutions
</h3>,
        <div
          className="evolution-chain-container"
>
          <div
                    className="evolution-chain-item first"
          >
                    <Evolution
                              getPokemonDetails={[Function]}
                              id={10}
                              image_path="/sprites/pokemon/small/10.png"
                              name="Caterpie"
                              types={
                                        Array [
                                                  "bug",
                                                ]
                              }
                    />
          </div>
          <div
                    className="evolution-chain-item second"
          >
                    <Evolution
                              getPokemonDetails={[Function]}
                              id={11}
                              image_path="/sprites/pokemon/small/11.png"
                              name="Metapod"
                              types={
                                        Array [
                                                  "bug",
                                                ]
                              }
                    />
          </div>
          <div
                    className="evolution-chain-item third"
          >
                    <Evolution
                              getPokemonDetails={[Function]}
                              id={12}
                              image_path="/sprites/pokemon/small/12.png"
                              name="Butterfree"
                              types={
                                        Array [
                                                  "bug",
                                                  "flying",
                                                ]
                              }
                    />
          </div>
</div>,
      ],
      "className": "evolution-chain-background",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Evolutions",
        },
        "ref": null,
        "rendered": "Evolutions",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="evolution-chain-item first"
>
              <Evolution
                            getPokemonDetails={[Function]}
                            id={10}
                            image_path="/sprites/pokemon/small/10.png"
                            name="Caterpie"
                            types={
                                          Array [
                                                        "bug",
                                                      ]
                            }
              />
</div>,
            <div
              className="evolution-chain-item second"
>
              <Evolution
                            getPokemonDetails={[Function]}
                            id={11}
                            image_path="/sprites/pokemon/small/11.png"
                            name="Metapod"
                            types={
                                          Array [
                                                        "bug",
                                                      ]
                            }
              />
</div>,
            <div
              className="evolution-chain-item third"
>
              <Evolution
                            getPokemonDetails={[Function]}
                            id={12}
                            image_path="/sprites/pokemon/small/12.png"
                            name="Butterfree"
                            types={
                                          Array [
                                                        "bug",
                                                        "flying",
                                                      ]
                            }
              />
</div>,
          ],
          "className": "evolution-chain-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Evolution
                  getPokemonDetails={[Function]}
                  id={10}
                  image_path="/sprites/pokemon/small/10.png"
                  name="Caterpie"
                  types={
                                    Array [
                                                      "bug",
                                                    ]
                  }
/>,
              ],
              "className": "evolution-chain-item first",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "class",
                "props": Object {
                  "getPokemonDetails": [Function],
                  "id": 10,
                  "image_path": "/sprites/pokemon/small/10.png",
                  "name": "Caterpie",
                  "types": Array [
                    "bug",
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Evolution
                  getPokemonDetails={[Function]}
                  id={11}
                  image_path="/sprites/pokemon/small/11.png"
                  name="Metapod"
                  types={
                                    Array [
                                                      "bug",
                                                    ]
                  }
/>,
              ],
              "className": "evolution-chain-item second",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "class",
                "props": Object {
                  "getPokemonDetails": [Function],
                  "id": 11,
                  "image_path": "/sprites/pokemon/small/11.png",
                  "name": "Metapod",
                  "types": Array [
                    "bug",
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Evolution
                  getPokemonDetails={[Function]}
                  id={12}
                  image_path="/sprites/pokemon/small/12.png"
                  name="Butterfree"
                  types={
                                    Array [
                                                      "bug",
                                                      "flying",
                                                    ]
                  }
/>,
              ],
              "className": "evolution-chain-item third",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "class",
                "props": Object {
                  "getPokemonDetails": [Function],
                  "id": 12,
                  "image_path": "/sprites/pokemon/small/12.png",
                  "name": "Butterfree",
                  "types": Array [
                    "bug",
                    "flying",
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Evolutions
</h3>,
          <div
            className="evolution-chain-container"
>
            <div
                        className="evolution-chain-item first"
            >
                        <Evolution
                                    getPokemonDetails={[Function]}
                                    id={10}
                                    image_path="/sprites/pokemon/small/10.png"
                                    name="Caterpie"
                                    types={
                                                Array [
                                                            "bug",
                                                          ]
                                    }
                        />
            </div>
            <div
                        className="evolution-chain-item second"
            >
                        <Evolution
                                    getPokemonDetails={[Function]}
                                    id={11}
                                    image_path="/sprites/pokemon/small/11.png"
                                    name="Metapod"
                                    types={
                                                Array [
                                                            "bug",
                                                          ]
                                    }
                        />
            </div>
            <div
                        className="evolution-chain-item third"
            >
                        <Evolution
                                    getPokemonDetails={[Function]}
                                    id={12}
                                    image_path="/sprites/pokemon/small/12.png"
                                    name="Butterfree"
                                    types={
                                                Array [
                                                            "bug",
                                                            "flying",
                                                          ]
                                    }
                        />
            </div>
</div>,
        ],
        "className": "evolution-chain-background",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Evolutions",
          },
          "ref": null,
          "rendered": "Evolutions",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="evolution-chain-item first"
>
                <Evolution
                                getPokemonDetails={[Function]}
                                id={10}
                                image_path="/sprites/pokemon/small/10.png"
                                name="Caterpie"
                                types={
                                                Array [
                                                                "bug",
                                                              ]
                                }
                />
</div>,
              <div
                className="evolution-chain-item second"
>
                <Evolution
                                getPokemonDetails={[Function]}
                                id={11}
                                image_path="/sprites/pokemon/small/11.png"
                                name="Metapod"
                                types={
                                                Array [
                                                                "bug",
                                                              ]
                                }
                />
</div>,
              <div
                className="evolution-chain-item third"
>
                <Evolution
                                getPokemonDetails={[Function]}
                                id={12}
                                image_path="/sprites/pokemon/small/12.png"
                                name="Butterfree"
                                types={
                                                Array [
                                                                "bug",
                                                                "flying",
                                                              ]
                                }
                />
</div>,
            ],
            "className": "evolution-chain-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Evolution
                    getPokemonDetails={[Function]}
                    id={10}
                    image_path="/sprites/pokemon/small/10.png"
                    name="Caterpie"
                    types={
                                        Array [
                                                            "bug",
                                                          ]
                    }
/>,
                ],
                "className": "evolution-chain-item first",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "class",
                  "props": Object {
                    "getPokemonDetails": [Function],
                    "id": 10,
                    "image_path": "/sprites/pokemon/small/10.png",
                    "name": "Caterpie",
                    "types": Array [
                      "bug",
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Evolution
                    getPokemonDetails={[Function]}
                    id={11}
                    image_path="/sprites/pokemon/small/11.png"
                    name="Metapod"
                    types={
                                        Array [
                                                            "bug",
                                                          ]
                    }
/>,
                ],
                "className": "evolution-chain-item second",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "class",
                  "props": Object {
                    "getPokemonDetails": [Function],
                    "id": 11,
                    "image_path": "/sprites/pokemon/small/11.png",
                    "name": "Metapod",
                    "types": Array [
                      "bug",
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Evolution
                    getPokemonDetails={[Function]}
                    id={12}
                    image_path="/sprites/pokemon/small/12.png"
                    name="Butterfree"
                    types={
                                        Array [
                                                            "bug",
                                                            "flying",
                                                          ]
                    }
/>,
                ],
                "className": "evolution-chain-item third",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "class",
                  "props": Object {
                    "getPokemonDetails": [Function],
                    "id": 12,
                    "image_path": "/sprites/pokemon/small/12.png",
                    "name": "Butterfree",
                    "types": Array [
                      "bug",
                      "flying",
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`EvolutionChain Component when there are no evolutions should match EvolutionChain component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EvolutionChain
    evolutions={
        Object {
            "1": null,
            "2": null,
            "3": null,
          }
    }
    getPokemonDetails={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          This pokemon doesn't evolve
</h3>,
        null,
      ],
      "className": "evolution-chain-background",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "This pokemon doesn't evolve",
        },
        "ref": null,
        "rendered": "This pokemon doesn't evolve",
        "type": "h3",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            This pokemon doesn't evolve
</h3>,
          null,
        ],
        "className": "evolution-chain-background",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "This pokemon doesn't evolve",
          },
          "ref": null,
          "rendered": "This pokemon doesn't evolve",
          "type": "h3",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
